RxSwift/Observables/Buffer.swift

```diff
@@ -136,3 +136,131 @@ final private class BufferTimeCountSink<Element, Observer: ObserverType>
         nextTimer.setDisposable(disposable)
     }
 }
+
+extension ObservableType {
+    
+    /**
+     Projects each element of an observable sequence info a buffer that's sent out when the trigger Observable emits a .next event or a .complete event, if the accumulated buffer's count
+     is over zero.
+     
+     - seealso: [buffer operator on reactivex.io](http://reactivex.io/documentation/operators/buffer.html)
+     
+     - parameter trigger: Observable that will act as a boundary trigger between each window.
+     - returns: An observable sequence of buffers.
+     */
+    public func buffer<TriggerElement>(trigger: Observable<TriggerElement>) -> Observable<[Element]> {
+        return BufferTrigger(source: self.asObservable(), trigger: trigger)
+    }
+    
+    /**
+     Projects each element of an observable sequence info a buffer that's sent out when the trigger Observable emits a .next event or a .complete event, if the accumulated buffer's count
+     is over zero.
+     
+     A useful real-world example of this is when you have a resource that needs to be updated depending on some asynchronous tasks and you want to accumulate all of those tasks's results
+     and just PUT/PATCH the remote resource once per batch of results instead of once per result.
+     
+     - seealso: [buffer operator on reactivex.io](http://reactivex.io/documentation/operators/buffer.html)
+     
+     - parameter debounce: Amount of time to debounce the source as the boundary trigger between each window.
+     - parameter scheduler: Scheduler to run debouncing on.
+     - returns: An observable sequence of buffers.
+     */
+    public func buffer(debounce: RxTimeInterval, scheduler: SchedulerType) -> Observable<[Element]> {
+        let shared = self.share()
+        return shared.buffer(trigger: shared.debounce(debounce, scheduler: scheduler))
```

Comment:
Is there a way to do this without share? I don't think we can introduce it this way because it inevitably makes the `Observable` shared downstream which could cause bugs and be an unexpected side-effect for consumers

Reply:
Ah, can you tell me more of what that could look like?

Reply:
You won't like this, but the right way is probably passing an optional `debounceTime` or something similar named to the `Producer` and `Sink` and have that logic sit there, without relying on `debounce` at all. 

In general it isn't a good practice for custom operators inside RxSwift to count on other operators as we prefer atomic units, as much as possible.

Reply:
Any suggestions about this one though?

Reply:
Sorry, as I asked for a suggestion the page hadn't refreshed so I missed the above.